apps/solana_app/solana_pub_key.c

```diff
@@ -0,0 +1,452 @@
+}
+
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+void solana_get_pub_keys(solana_query_t *query) {
+  char wallet_name[NAME_SIZE] = "";
+  uint8_t seed[64] = {0};
+
+  const pb_size_t which_request = query->which_request;
+  solana_get_public_keys_intiate_request_t *init_req = NULL;
+  pb_size_t which_response = SOLANA_RESULT_COMMON_ERROR_TAG;
+
+  if (SOLANA_QUERY_GET_PUBLIC_KEYS_TAG == which_request) {
+    which_response = SOLANA_RESULT_GET_PUBLIC_KEYS_TAG;
+    init_req = &query->get_public_keys.initiate;
+  } else {
+    which_response = SOLANA_RESULT_GET_USER_VERIFIED_PUBLIC_KEY_TAG;
+    init_req = &query->get_user_verified_public_key.initiate;
+  }
+
+  uint8_t public_keys[sizeof(init_req->derivation_paths) /
+                      sizeof(solana_get_public_keys_derivation_path_t)]
+                     [SOLANA_PUB_KEY_SIZE] = {0};
+
+  if (!check_which_request(query,
+                           SOLANA_GET_PUBLIC_KEYS_REQUEST_INITIATE_TAG) ||
+      !validate_request_data(&query->get_public_keys, which_request) ||
+      !get_wallet_name_by_id(query->get_public_keys.initiate.wallet_id,
+                             (uint8_t *)wallet_name,
+                             solana_send_error)) {
+    return;
+  }
+
+  // Take user consent to export public key for the wallet
+  if (!get_user_consent(which_request, wallet_name)) {
+    return;
+  }
+
+  set_app_flow_status(SOLANA_GET_PUBLIC_KEYS_STATUS_CONFIRM);
+
+  if (!reconstruct_seed(query->get_public_keys.initiate.wallet_id,
+                        &seed[0],
+                        solana_send_error)) {
+    memzero(seed, sizeof(seed));
+    return;
+  }
+
+  set_app_flow_status(SOLANA_GET_PUBLIC_KEYS_STATUS_SEED_GENERATED);
+  delay_scr_init(ui_text_processing, DELAY_SHORT);
+
+  bool status = fill_public_keys(init_req->derivation_paths,
+                                 seed,
+                                 public_keys,
+                                 init_req->derivation_paths_count);
+
+  // Clear seed as soon as it is not needed
+  memzero(seed, sizeof(seed));
+
+  if (!status) {
+    // send unknown error; do not know failure reason
+    solana_send_error(ERROR_COMMON_ERROR_UNKNOWN_ERROR_TAG, 1);
+    return;
+  }
+
+  if (SOLANA_QUERY_GET_USER_VERIFIED_PUBLIC_KEY_TAG == which_request) {
+    char address[100] = "";
+
+    size_t public_key_size = sizeof(address);
+    if (!b58enc(address, &public_key_size, (char *)(public_keys[0]), 32)) {
```

Comment:
is the explicit typecasting of 2nd parameter (`(char *)(public_keys[0])`) necessary?

Reply:
ah, b58enc accepts `const void*`, and casts it into  `uint8_t *`.
however it's done so in the previous implementation
`  b58enc(receive_transaction_data.solana_address,
         &public_key_size,
         (char *)(node.public_key + 1),
         32)`

Reply:
okay. check what is best. explicit casting should not be needed IMHO. near is a good reference for this. because both (Near and Solana) use the same Elliptic curve.